packages/patterns/src/keys/keycollection-operators.js

```diff
@@ -0,0 +1,226 @@
+/** @typedef {import('../types').KeyComparison} KeyComparison */
+/** @typedef {import('../types').KeyCompare} KeyCompare */
+/** @typedef {import('../types').FullCompare} FullCompare */
+/** @typedef {import('../types').KeyCollection} KeyCollection */
+
+const { quote: q, Fail } = assert;
+
+/**
+ * Refines a sequence of entries that is already sorted over its keys by the
+ * `rankCompare` preorder, where there may be internal runs tied for the same
+ * rank, into an iterable that resolves those ties using `fullCompare`.
+ *
+ * @template [V=unknown]
+ * @param {Iterable<[Key, V]>} rankSorted
+ * @param {RankCompare} rankCompare
+ * @param {FullCompare} fullCompare
+ * @returns {IterableIterator<[Key, V]>}
+ */
+const generateFullSortedEntries = (rankSorted, rankCompare, fullCompare) => {
+  /** @type {Array<[Key, V]>} */
+  let arrEntries;
+  if (Array.isArray(rankSorted)) {
+    arrEntries = rankSorted;
+  } else {
+    arrEntries = harden([...rankSorted]);
+  }
+  assertRankSorted(arrEntries, rankCompare);
+  const { length } = arrEntries;
+  let i = 0;
+  let sameRankIterator;
+  return makeIterator(() => {
+    if (sameRankIterator) {
+      const result = sameRankIterator.next();
+      if (!result.done) {
+        return result;
+      }
+      sameRankIterator = undefined;
+    }
+    if (i < length) {
+      const entry = arrEntries[i];
+      // Look ahead for same-rank ties.
+      let j = i + 1;
+      while (j < length && rankCompare(entry[0], arrEntries[j][0]) === 0) {
+        j += 1;
+      }
+      if (j === i + 1) {
+        // No ties found.
+        i = j;
+        return harden({ done: false, value: entry });
+      }
+      const ties = arrEntries.slice(i, j);
+      i = j;
+
+      // Sort the ties by `fullCompare`, enforce key uniqueness, and delegate to
+      // a sub-iterator.
+      const sortedTies = sortByRank(ties, fullCompare);
+      for (let k = 1; k < sortedTies.length; k += 1) {
+        const [key0] = sortedTies[k - 1];
+        const [key1] = sortedTies[k];
+        Math.sign(fullCompare(key0, key1)) ||
+          Fail`Duplicate entry key: ${key0}`;
+      }
+      sameRankIterator = makeArrayIterator(sortedTies);
+      return sameRankIterator.next();
+    }
+    return harden({ done: true, value: undefined });
+  });
+};
+harden(generateFullSortedEntries);
+
+/**
+ * Returns an iterator that merges reverse-rank-sorted [key, value] entries of
+ * two KeyCollections into a reverse-full-sorted [key, value1, value2] entries
+ * by the key they have in common, representing the value for an absent entry in
+ * either collection as `absentValue`.
+ *
+ * @template [C=KeyCollection]
+ * @template [V=unknown]
+ * @param {C} c1
+ * @param {C} c2
+ * @param {(collection: C) => Iterable<[Key, V]>} getEntries
+ * @param {any} absentValue
+ * @returns {IterableIterator<[Key, V | absentValue, V | absentValue]>}
+ */
+export const generateCollectionPairEntries = (
```

Comment:
Likewise, shouldn't this replace both set's and bag's `merge` functions?

Reply:
Yes! ...but maybe not here.

Reply:
Understood. Resolving.